internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,490 @@
+package authorization
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"maps"
+	"sort"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	apimachyaml "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	rbacinternal "k8s.io/kubernetes/pkg/apis/rbac"
+	rbacv1helpers "k8s.io/kubernetes/pkg/apis/rbac/v1"
+	rbacregistry "k8s.io/kubernetes/pkg/registry/rbac"
+	"k8s.io/kubernetes/pkg/registry/rbac/validation"
+	rbac "k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type PreAuthorizer interface {
+	PreAuthorize(ctx context.Context, manifestManager user.Info, manifestReader io.Reader) (map[string][]rbacv1.PolicyRule, error)
+}
+
+var (
+	collectionVerbs = []string{"list", "watch", "create"}
+	objectVerbs     = []string{"get", "patch", "update", "delete"}
+)
+
+type rbacPreAuthorizer struct {
+	authorizer   authorizer.Authorizer
+	ruleResolver validation.AuthorizationRuleResolver
+	restMapper   meta.RESTMapper
+}
+
+func NewRBACPreAuthorizer(cl client.Client) PreAuthorizer {
+	return &rbacPreAuthorizer{
+		authorizer:   newRBACAuthorizer(cl),
+		ruleResolver: newRBACRulesResolver(cl),
+		restMapper:   cl.RESTMapper(),
+	}
+}
+
+// PreAuthorize validates whether the current user/request satisfies the necessary permissions
+// as defined by the RBAC policy. It examines the user’s roles, resource identifiers, and
+// the intended action to determine if the operation is allowed.
+//
+// Return Value:
+//   - nil: indicates that the authorization check passed and the operation is permitted.
+//   - non-nil error: indicates that the authorization failed (either due to insufficient permissions
+//     or an error encountered during the check), the error provides a slice of several failures at once.
```

Comment:
Non-nil errors are _just_ when there were issues evaluating the permissions, right? If we successfully evaluate permissions and determine that some rules are missing, then we'll return a `nil` error with some missing rules.

I would also say that the returned error _may_ include multiple failures encountered during rule evaluation (but it also may not, for example if there was a failure decoding the manifest).